src/app/store/observeMiddleware.ts

```diff
@@ -0,0 +1,76 @@
+import { deepEqual } from 'fast-equals';
+import { Dispatch, isAction, Middleware, MiddlewareAPI } from 'redux';
+import { createAction, createCustomAction, isActionOf } from 'typesafe-actions';
+import { RootState } from './types';
+
+interface StoreObserver<T> {
+  id: string;
+  /**
+   * Function to create "something" which will be used to determine if the side effect should run.
+   * deepEqual is used for the comparison, so use primitives, shallow objects or arrays to keep
+   * performance reasonable.
+   */
+  getObserved: (rootState: RootState) => T;
+  /**
+   * Runs the side effect providing both the previous and current version of the derrived state.
+   */
+  sideEffect: (states: { previous: T; current: T }) => void;
+}
+
+// Need to user a higher order function to get the correct typings and inference, it will now correctly
+// type the value of T based on what the getObserved returns
+export const observe = <T>(storeObserver: StoreObserver<T>) =>
+  createCustomAction('observer/OBSERVE', (storeObserver: StoreObserver<T>) => ({ storeObserver }))(
+    storeObserver,
+  );
+export const unobserve = createAction('observer/UNOBSERVE')<string>();
+export const clearObservers = createAction('observer/CLEAR_OBSERVERS')();
+
+function isObserveAction(action: unknown): action is ReturnType<typeof observe> {
+  return (
+    typeof action === 'object' &&
+    action !== null &&
+    'type' in action &&
+    action.type === 'observer/OBSERVE'
+  );
+}
```

Comment:
I *think* you can use `isActionOf(observe, action)`?

Reply:
I tried initially but it didn't seem to work with the custom action. 

I'll have another go when I get back to the PR and report back the type issues I see (if they still exist)

Reply:
Yup, that works. Just me grappling with the library initially.

Reply:
Ah I have run into the issue again while testing. While the type system is fine, the validation system for typesafe actions throws

```
validation.ts:40 Uncaught (in promise) Error: Argument contains array with invalid element at index 0, it should be an action-creator instance from "typesafe-actions"
    at o (validation.ts:40:11)
    at Array.forEach (<anonymous>)
    at push../node_modules/.pnpm/typesafe-actions@5.1.0/node_modules/typesafe-actions/dist/typesafe-actions.umd.production.js.n.isActionOf (is-action-of.ts:45:18)
    at Object.dispatch (observeMiddleware.ts:46:19)
    at Object.dispatch (<anonymous>:6:7384)
    at Index.tsx:72:9
```